modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java

```diff
@@ -428,7 +431,8 @@ private CompletableFuture<AsyncSqlCursor<List<Object>>> querySingle0(
                                 InternalTransaction implicitTx = implicitTxRequired ? txManager.begin() : null;
 
                                 BaseQueryContext enrichedContext =
-                                        implicitTxRequired ? ctx.toBuilder().transaction(implicitTx).build() : ctx;
+                                        implicitTxRequired ? ctx.toBuilder().transaction(implicitTx)
+                                                .transactionId(distributedSqlDisabled ? null : implicitTx.id()).build() : ctx;
```

Comment:
As I understand, you suggest to set `transactionId` mandatory.
But such change will require to change condition order in scan nodes.
```
if (readOnlyTx) {
    pub = physTable.scan(...
} else if (context().transactionId() != null) {
    pub = physTable.scan(...
} else {
    // TODO IGNITE-17952 This block should be removed.
    // Workaround to make RW scan work from tx coordinator.
    pub = physTable.scan(partId, context().transaction());
}
```
to something like this
```
if (readOnlyTx) {
    pub = physTable.scan(...
} else if (context.transaction() != null) {
    // TODO IGNITE-17952 This block should be removed.
    // Workaround to make RW scan work from tx coordinator.
    pub = physTable.scan(partId, context().transaction());
} else {
    pub = physTable.scan(...
} 
```
otherwise, we will try to do an RW scan on the local host because DISTRIBUTED_TRAIT is currently disabled and we send all plan fragments to the local host.

Is it ok?

Reply:
After some thought, I decided that this makes the code look cleaner and easier to make the changes needed to enable distributive SQL (in IGNITE-17952). So I decided to follow your suggestion and made `txId` mandatory, thanks.